dbms/src/Interpreters/JoinHashTable.h

```diff
@@ -0,0 +1,155 @@
+struct RowRefList : RowRef
+{
+    RowRefList * next = nullptr;
+
+    RowRefList() = default;
+    RowRefList(const Block * block_, size_t row_num_)
+        : RowRef(block_, row_num_)
+    {}
+};
+
+/** Depending on template parameter, adds or doesn't add a flag, that element was used (row was joined).
+      * For implementation of RIGHT and FULL JOINs.
+      * NOTE: It is possible to store the flag in one bit of pointer to block or row_num. It seems not reasonable, because memory saving is minimal.
+      */
+template <bool enable, typename Base>
+struct WithUsedFlag;
+
+template <typename Base>
+struct WithUsedFlag<true, Base> : Base
+{
+    mutable std::atomic<bool> used{};
+    using Base::Base;
+    using Base_t = Base;
+    void setUsed() const { used.store(true, std::memory_order_relaxed); } /// Could be set simultaneously from different threads.
+    bool getUsed() const { return used.load(std::memory_order_relaxed); }
+};
+
+template <typename Base>
+struct WithUsedFlag<false, Base> : Base
+{
+    using Base::Base;
+    using Base_t = Base;
+    void setUsed() const {}
+    bool getUsed() const { return true; }
+};
+
+
+/// Different types of keys for maps.
+#define APPLY_FOR_JOIN_VARIANTS(M) \
+    M(key8)                        \
+    M(key16)                       \
+    M(key32)                       \
+    M(key64)                       \
+    M(key_string)                  \
+    M(key_strbinpadding)           \
+    M(key_strbin)                  \
+    M(key_fixed_string)            \
+    M(keys128)                     \
+    M(keys256)                     \
+    M(serialized)
+
+enum class JoinType
+{
+    EMPTY,
+    CROSS,
+#define M(NAME) NAME,
+    APPLY_FOR_JOIN_VARIANTS(M)
+#undef M
+};
+
+
+/** Different data structures, that are used to perform JOIN.
+      */
+template <typename Mapped>
+struct ConcurrentMapsTemplate
+{
+    std::unique_ptr<ConcurrentHashMap<UInt8, Mapped, TrivialHash, HashTableFixedGrower<8>>> key8;
+    std::unique_ptr<ConcurrentHashMap<UInt16, Mapped, TrivialHash, HashTableFixedGrower<16>>> key16;
+    std::unique_ptr<ConcurrentHashMap<UInt32, Mapped, HashCRC32<UInt32>>> key32;
+    std::unique_ptr<ConcurrentHashMap<UInt64, Mapped, HashCRC32<UInt64>>> key64;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> key_string;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> key_strbinpadding;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> key_strbin;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> key_fixed_string;
+    std::unique_ptr<ConcurrentHashMap<UInt128, Mapped, HashCRC32<UInt128>>> keys128;
+    std::unique_ptr<ConcurrentHashMap<UInt256, Mapped, HashCRC32<UInt256>>> keys256;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> serialized;
+    // TODO: add more cases like Aggregator
+};
+
+template <typename Mapped>
+struct MapsTemplate
```

Comment:
The code style between `MapsTemplate` and `ConcurrentMapsTemplate` is different. I am ok with these two styles but they should be the same.

Reply:
Updated.